stl/inc/flat_map

```diff
@@ -0,0 +1,953 @@
+                  || (constructible_from<key_type, _K&&> && _Compare_is_transparent<key_compare>) )
+    {
+        return _Emplace_hint<false, false>(_Position, _STD forward<_K>(_Key), _STD forward<_Args_t>(_Args)...);
+    }
+
+    template <class _M, class _K>
+    pair<iterator, bool> insert_or_assign(_K&& _Key, _M&& _Obj)
+        requires assignable_from<mapped_type&, _M&&> && constructible_from<mapped_type, _M&&>
+              && (same_as<remove_cvref_t<_K &&>, key_type>
+                  || (constructible_from<key_type, _K&&> && _Compare_is_transparent<key_compare>) )
+    {
+        auto _Res = try_emplace(_STD forward<_K>(_Key), _STD forward<_M>(_Obj));
+        if (_Res.second) {
+            // Insertion took place
+            return _Res;
+        } else {
+            // Already exists
+            *(_Res.first._Mapped_it) = _STD forward<_M>(_Obj);
+            return _Res;
+        }
+    }
+
+    template <class _M, class _K>
+    iterator insert_or_assign(const_iterator _Position, _K&& _Key, _M&& _Obj)
+        requires assignable_from<mapped_type&, _M&&> && constructible_from<mapped_type, _M&&>
+              && (same_as<remove_cvref_t<_K &&>, key_type>
+                  || (constructible_from<key_type, _K&&> && _Compare_is_transparent<key_compare>) )
+    {
+        return _Emplace_hint<false, true>(_Position, _STD forward<_K>(_Key), _STD forward<_M>(_Obj));
+    }
+
+    iterator erase(iterator _Position) {
+        return erase(static_cast<const_iterator>(_Position));
+    }
+
+    iterator erase(const_iterator _Position) {
+        _Clear_flat_map_scope_guard _Guard{this};
+        auto _Key_it      = _Data.keys.erase(_Position._Key_it);
+        auto _Val_it      = _Data.values.erase(_Position._Mapped_it);
+        _Guard._Clearable = nullptr;
+        return iterator{_STD move(_Key_it), _STD move(_Val_it)};
+    }
+
+    iterator erase(const_iterator _First, const_iterator _Last) {
+        _Clear_flat_map_scope_guard _Guard{this};
+        auto _Key_it      = _Data.keys.erase(_First._Key_it, _Last._Key_it);
+        auto _Val_it      = _Data.values.erase(_First._Mapped_it, _Last._Mapped_it);
+        _Guard._Clearable = nullptr;
+        return iterator{_STD move(_Key_it), _STD move(_Val_it)};
+    }
+
+    template <class _K>
+    size_type erase(_K&& _Key)
+        requires convertible_to<_K&&, const key_type&>
+              || (_Compare_is_transparent<key_compare> && !convertible_to<_K &&, iterator>
+                  && !convertible_to<_K &&, const_iterator>)
+    {
+        const_iterator _Pos_begin = lower_bound(_STD forward<_K>(_Key));
+        const_iterator _Pos_end   = upper_bound(_STD forward<_K>(_Key));
+        size_type _Count          = _Pos_end - _Pos_begin;
+        erase(_Pos_begin, _Pos_end);
+        return _Count;
+    }
+
+    containers extract() && {
+        _Clear_flat_map_scope_guard _Guard{this};
+        return _STD move(_Data);

```

Comment:
The `move` is needed in this return statement, right?

Reply:
> The move is needed in this return statement, right?

Yes, this is needed.

And by the way, I've made a mistake in `flat_set` receltly ~ trying to elide the `clear()` call if move-ctor is `noexcept`. This is invalid, as even a successful `noexcept` move constrction cannot guarantee that the original container is emptied (for example, a container is *allowed* just to copy the data for move-ctor).